### DISCUSSION TITLE
Display image over at most X columns

### DISCUSSION THREAD

**Discussion by swsnr** [kovidgoyal/kitty#6201] [Q&A] [answered]
Using the graphics protocol how can I display an image over *at most* X columns, i.e. have kitty downscale images wider than the given amount of columns X, but render at original size if the image is smaller than X columns? Like the c control code, but without upscaling?

Alternatively, I can scale images myself, but then I'd presumably need to know kitty's column width/cell size in pixels to correctly calculate the pixel width corresponding to the desired amount of columns.  How could I get this information?

The idea here is to have images downscaled to a fix a text wrapped at a given column limit in [mdcat](https://github.com/swsnr/mdcat) , but never upscale small images (like the wide-spread README badges on GitHub) to fill an entire line.

PS: Ideally I'd love to have a solution which doesn't involve requesting information from the terminal by means of escape codes and reading the terminal responses, because that's quite complicated and requires that the rendering code actually runs on the target terminal.  However, I'd love to be able to pipe rendering output into a text file and later `cat` that file to reproduce the rendering, i.e. render "offline" in a way.

**Comment by kovidgoyal** [comment, marked as answer]
There is no such operation. The graphics protocol page includes  sample code in several languages to get the cell dimensions. https://sw.kovidgoyal.net/kitty/graphics-protocol/#getting-the-window-size  simply dive the screen size in pixels by the screen size in cells.

**Reply by swsnr** [reply]
I'm aware if this and I actually use the pixel size already, but I wasn't aware that the pixel width directly relates to the column size.

I thought the window size would include eg borders or padding, or "half cells" if the window size doesn't match the column width perfectly?

So you're saying that the pixel width I get out of that ioctl is always divisible by by the column width, without any remainder?

**Comment by kovidgoyal** [comment]
yes, that's what I am saying.

**Reply by swsnr** [reply]
Great thank you very much!
